File: DMOPC '21 Contest 6 P3 - An Art Problem.cpp
// https://dmoj.ca/problem/dmopc21c6p3
// BFS flood fill, multiple starting points
// sort to start by filling with the smaller numbers

#include <bits/stdc++.h>

using namespace std;

bool sort3rd(const tuple<int, int, int, int> &a,
              const tuple<int, int, int, int> &b) {
    return (get<2>(a) < get<2>(b));
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int ROWS, COLS, K;
    cin >> ROWS >> COLS >> K;

    vector<vector<int>> graph(ROWS, vector<int>(COLS));
    for (int i = 0; i < ROWS; ++i)
        for (int j = 0; j < COLS; ++j)
            cin >> graph[i][j];

    array<pair<int, int>, 4> dir = {{{0, 1}, {1, 0}, {-1, 0}, {0, -1}}};

    vector<tuple<int, int, int, int>> q;
    for (int i = 0; i < ROWS; ++i)
        for (int j = 0; j < COLS; ++j)
            if (graph[i][j])
                q.emplace_back(i, j, graph[i][j], 1);

    sort(q.begin(), q.end(), sort3rd);  // sort by color, smaller colors have priority

    deque<tuple<int, int, int, int>> dq(q.begin(), q.end());
    while (!dq.empty()) {
        auto [row, col, color, time] = dq.front();
        dq.pop_front();
        if (time > K)
            break;
        for (auto [dr, dc]: dir) {
            int new_r = row + dr;
            int new_c = col + dc;
            if (0 <= new_r && new_r < ROWS && 0 <= new_c && new_c < COLS && !graph[new_r][new_c]) {
                graph[new_r][new_c] = color;
                dq.emplace_back(new_r, new_c, color, time + 1);
            }
        }
    }

    for (auto &row: graph) {
        for (auto &cell: row)
            cout << cell << ' ';
        cout << '\n';
    }

    return 0;
}

File: DMOPC '21 Contest 6 P3 - An Art Problem.py
# TLE, check c++ version for code that passes with exact same logic

# https://dmoj.ca/problem/dmopc21c6p3
# BFS flood fill, multiple starting points
# sort to start by filling with the smaller numbers

from collections import deque
import sys

input = sys.stdin.readline

ROWS, COLS, K = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(ROWS)]
dir = [(0, 1), (1, 0), (-1, 0), (0, -1)]

q = []
for i in range(ROWS):
    for j in range(COLS):
        if graph[i][j]:
            q.append((i, j, graph[i][j], 1))

q.sort(key=lambda x: x[2])
q = deque(q)
while q:
    row, col, color, time = q.popleft()
    if time > K:
        break
    for dr, dc in dir:
        new_r = row + dr
        new_c = col + dc
        if 0 <= new_r < ROWS and 0 <= new_c < COLS and not graph[new_r][new_c]:
            graph[new_r][new_c] = color
            q.append((new_r, new_c, color, time + 1))

for row in graph:
    print(*row)
